### PR TITLE
Finish and Test the simultaneous connect problem in libp2p peers

### PR DESCRIPTION
**Pull request by aarshkshah1992**
Closes https://github.com/libp2p/go-libp2p-noise/issues/70.

### TODO
 - [ ] Update Deps and then merge.

### CODE REVIEW

File: config/security.go [hunk 1]
```diff
@@ -50,13 +50,13 @@ func SecurityConstructor(security interface{}) (SecC, error) {
...
 
-func makeInsecureTransport(id peer.ID, privKey crypto.PrivKey) sec.SecureTransport {
+func makeInsecureTransport(id peer.ID, privKey crypto.PrivKey) sec.SecureMuxer {
 	secMuxer := new(csms.SSMuxer)
 	secMuxer.AddTransport(insecure.ID, insecure.NewWithIdentity(id, privKey))
 	return secMuxer
 }
 
-func makeSecurityTransport(h host.Host, tpts []MsSecC) (sec.SecureTransport, error) {
+func makeSecurityTransport(h host.Host, tpts []MsSecC) (sec.SecureMuxer, error) {
```
Comment (marten-seemann) [inline config/security.go:59]: `makeSecurityMuxer`?

**Comment by aarshkshah1992** [comment]
cc @vyzo Please can you give me a review ?

**Review by vyzo — APPROVED** [review]

File: libp2p_test.go [hunk 2]
```diff
@@ -192,3 +196,52 @@ func TestChainOptions(t *testing.T) {
...
+
+	wg := new(sync.WaitGroup)
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for i := 0; i < 20; i++ {
+			require.NoError(t, h1.Connect(context.Background(), h2Info))
+			require.NoError(t, h1.Network().ClosePeer(h2.ID()))
+		}
```
Comment (raulk) [inline libp2p_test.go:233]: Are 20 iterations sufficient to reliably trigger simopen?

Comment (aarshkshah1992) [reply libp2p_test.go:233]: @raulk Pretty much.

**Here's a run of the same test against `v0.13.0`**(previous release, did not have hole punching), and it fails repeatedly:

```
aarshshah@Aarshs-MacBook-Pro go-libp2p % go test -run TestTcpSimultaneousConnect -count 10
--- FAIL: TestTcpSimultaneousConnect (0.28s)
    libp2p_test.go:231: 
        	Error Trace:	libp2p_test.go:231
        	            				asm_amd64.s:1374
        	Error:      	Received unexpected error:
        	            	failed to dial QmZsP5W7ceKhynLcEtQpHXRTNTfd7i9YnBsnndwvmi5p3t: all dials failed
        	            	  * [/ip4/192.168.43.87/tcp/50093] failed to negotiate security protocol: error reading handshake message: noise: message is too short
        	            	  * [/ip4/127.0.0.1/tcp/50093] failed to negotiate security protocol: error reading handshake message: noise: message is too short
        	Test:       	TestTcpSimultaneousConnect
    libp2p_test.go:241: 
        	Error Trace:	libp2p_test.go:241
        	            				asm_amd64.s:1374
        	Error:      	Received unexpected error:
        	            	failed to dial QmcMLKDi4aSWSkMk1W7pHY5uFiB6Wi7yKVzgVTY6oWL4an: all dials failed
        	            	  * [/ip4/192.168.43.87/tcp/50092] failed to negotiate security protocol: error reading handshake message: noise: message is too short
        	            	  * [/ip4/127.0.0.1/tcp/50092] failed to negotiate security protocol: error reading handshake message: noise: message is too short
        	Test:       	TestTcpSimultaneousConnect
--- FAIL: TestTcpSimultaneousConnect (0.23s)
    libp2p_test.go:241: 
        	Error Trace:	libp2p_test.go:241
        	            				asm_amd64.s:1374
        	Error:      	Received unexpected error:
        	            	failed to dial Qmed9XqtjKjxEEoJi12PwKAf2NW35qkA6L2jffMEaodGQ2: all dials failed
        	            	  * [/ip4/192.168.43.87/tcp/50095] failed to negotiate security protocol: error reading handshake message: noise: message is too short
        	            	  * [/ip4/127.0.0.1/tcp/50095] failed to negotiate security protocol: error reading handshake message: noise: message is too short
        	Test:       	TestTcpSimultaneousConnect
    libp2p_test.go:231: 
        	Error Trace:	libp2p_test.go:231
        	            				asm
        	            				
  ...........................
  ...........................
  
  FAIL
exit status 1
FAIL	github.com/libp2p/go-libp2p	3.496s      	            				
```


**Now, here's the same run with simultaneous connect feature:**

```
aarshshah@Aarshs-MacBook-Pro go-libp2p % go test -run TestTcpSimultaneousConnect -count 10
go: downloading github.com/libp2p/go-libp2p-core v0.8.2
go: downloading github.com/libp2p/go-libp2p-transport-upgrader v0.4.2
go: downloading github.com/libp2p/go-libp2p-swarm v0.4.2
PASS
ok  	github.com/libp2p/go-libp2p	5.609s

```

**Pull request merged**